WV-2126 Right Click Menu

jwhurley1 opened the pull request:
## Description

Created a right click/context menu that gives the user some quick shortcuts to some commonly used features.

![image](https://user-images.githubusercontent.com/95651563/153051242-5434bfa2-3dfd-4e9d-9546-2537ae717068.png)

## How to Test
1. Run `git checkout wv-2126`
2. Run `npm ci && npm run build && npm start` to update your local environment and start up the local server.
3. Go to [localhost:3000](http://localhost:3000)
4. Right click anywhere on the map. Your OS/Browser context menu should not appear at all.  Instead, a semi-transparent dark gray box with white text and icons should appear, as shown in the screenshot above.
5. The following menu items should show and execute the following actions:

- Copy to Clipboard
  - Copies the coordinates that were collected during the right click directly to the user's clipboard
  - Flashes a tooltip above the right click menu confirming the coordinates have been copied
- Add a Place Marker
  - Adds a place marker at the coordinates that were collected during the right click
  - Appears at the top-left of the context menu
- Measure Distance in [unit]
  - Activates the distance measuring tool and goes into distance measuring "mode"
  - [unit] updates when "Change Units to..." is clicked in context menu
- Measure Area in [unit]
  - Activates the area measuring tool and goes into area measuring "mode"
  - [unit] updates when "Change Units to..." is clicked in context menu
- Remove Measurements
  - Clears all measurements on the map
- Change Units to [unit]
  - Toggles between "km" and "mi"
  - Little toggle icon is "off" when in "km"
  - Little toggle icon is "on" when in "mi"

minniewong reviewed (changes requested):
Super cool to see this in action! 
Couple of notes: 
1. The coordinates are backwards. E.g Using right click to get coordinates for Washington DC is this: -77.02684566085335,38.846113208612174, but actually it should be reversed to 38.846113208612174,-77.02684566085335
2. Thoughts on listing the actual coordinates (like google maps does) as the first option? The coordinates should reflect the coordinate format that has been set in the lower right corner.
3. I'm getting phantom "Copy Coordinates to Clipboard" tooltips. It happens when I try each option in the menu then when I get to "Measure Distance in km", the phantom tooltip appears and doesn't go away. 
![Screen Shot 2022-02-08 at 1 48 53 PM](https://user-images.githubusercontent.com/10747923/153055562-2f1655e4-16c7-4ec0-aa05-ceedd83cd7d4.png)

jasontk19 reviewed (changes requested):
1. Opening the context menu near the right side or bottom of the screen causes it to bunch up or go off-screen entirely (see screenshots below).  We will need to do some math to determine which direction to open the menu, or consider using an existing context-menu library that will already do this for us like: https://github.com/vkbansal/react-contextmenu
2. In general the amount of words and icons make the menu look really busy.  The icons are nice but I wonder if it would look cleaner without them.  Also I think some of the options text be shortened:
* Add place marker
* Copy coordinates
* Measure distance
* Measure area
* etc...
3.  Menu choices should be grouped by action with a divider between groups, for example:
![Screen Shot 2022-02-08 at 1 28 41 PM](https://user-images.githubusercontent.com/1480000/153070304-5dde215d-09ce-41ce-b0cb-4ef1244ef2a4.png)
4. The menu should close when any action is taken, right now it seems to only close when a measurement is initiated.
5. Related to number 2, I think the padding could be dropped from 15px to 5px for a more compact menu. 

![Screen Shot 2022-02-08 at 1 20 43 PM](https://user-images.githubusercontent.com/1480000/153070797-25fcfef1-f836-4963-9f89-563d04c4cbb8.png)
![Screen Shot 2022-02-08 at 1 20 34 PM](https://user-images.githubusercontent.com/1480000/153070808-4823f4e0-847d-48eb-b3c6-cc76cbf922a1.png)

jwhurley1 commented:
@minniewong 
> 1. The coordinates are backwards. E.g Using right click to get coordinates for Washington DC is this: -77.02684566085335,38.846113208612174, but actually it should be reversed to 38.846113208612174,-77.02684566085335

Fixed in https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31

> 2. Thoughts on listing the actual coordinates (like google maps does) as the first option? The coordinates should reflect the coordinate format that has been set in the lower right corner.

In https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31,  I listed the coordinates at the top of the right-click menu, but also included "Copy Coordinates" right below them.  Clicking on either the coordinates or "Copy Coordinates" copies the coordinates to the clipboard.  Just having the coordinates on there didn't seem clear to me that you could click on them to copy which is why I put "Copy Coordinates" below.  Obviously redundant, but wanted to put both on there to see what you think.  **Maybe we could put the copy icon next to the coordinates like in the original post and nix the "Copy Coordinates"?**

> 3. I'm getting phantom "Copy Coordinates to Clipboard" tooltips. It happens when I try each option in the menu then when I get to "Measure Distance in km", the phantom tooltip appears and doesn't go away even after making a measurement.

Fixed in https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31

jwhurley1 commented:
@jasontk19 
> 1. Opening the context menu near the right side or bottom of the screen causes it to bunch up or go off-screen entirely (see screenshots below).  We will need to do some math to determine which direction to open the menu, or consider using an existing context-menu library that will already do this for us like: https://github.com/vkbansal/react-contextmenu

Thanks for sharing this.  I have not implemented this for https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31, but I will check it out!

> 2. In general the amount of words and icons make the menu look really busy.  The icons are nice but I wonder if it would look cleaner without them.  Also I think some of the options text be shortened:
> 
> * Add place marker
> * Copy coordinates
> * Measure distance
> * Measure area
> * etc...

I removed the icons and shortened the menu items.  Fixed in https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31

> 3. Menu choices should be grouped by action with a divider between groups, for example:

Nice idea!  I put a line after "Add Place Marker".  That was the only line I put in there.  "Change Units" seemed relevant to the measurements so I left those all together.

![image](https://user-images.githubusercontent.com/95651563/153305728-52462eec-badc-4179-8980-6f501e2f0afd.png)

> 4. The menu should close when any action is taken, right now it seems to only close when a measurement is initiated.

Addressed in https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31.  The only goofy thing is when clicking the coordinates.  I put a `setTimeout` of 1.5 seconds on `copyCoordsToClipboard()` function which keeps the context menu open for 1.5 seconds to display the "Copied to clipboard!" message then closes the context menu.  Thoughts?

> 5. Related to number 2, I think the padding could be dropped from 15px to 5px for a more compact menu.

Addressed in https://github.com/nasa-gibs/worldview/pull/3810/commits/bff35ed1e829b82503ab41e5fce5df0ecac51d31.

> 6. This may address some of the above: consider re-using our `.modal-dialog` styles for this menu for consistency with other menus:

Thanks for sharing these.  I tried using the `.modal-dialog` classes but was getting hung up on trying to get it to work properly.  I can revisit tomorrow  I ended up sticking with my `.context-menu` class for now and just ported the some of those styles over to it.  I think ideally I'll want to use the `.modal-dialog` stuff so everything is shared and can be updated in one place in the future.

minniewong reviewed (approved):
These updates look really good! 👍

jasontk19 commented:
Changes are looking good!  Some additional feedback:

1. I think the timeout on the copy clipboard tooltip is fine, I think you could even shorten it a bit and it would be OK
2. I noticed you are using `navigator.clipboard.writeText()` to do the copying, consider using the `copy-to-clipboard` lib since we are already including it in the project and it may offer a broader range of browser and device support.
3. There is an error that occurs when you copy coordinates from the location marker tooltip/modal:
![Screen Shot 2022-02-10 at 9 05 13 AM](https://user-images.githubusercontent.com/1480000/153449000-e5964621-4554-4941-9ccb-f312796efd50.png)
![Screen Shot 2022-02-10 at 9 05 20 AM](https://user-images.githubusercontent.com/1480000/153449014-4bea58be-6a71-4be6-8672-a0b53399e816.png)
It looks like that component is making assumptions about what elements are available on screen, this is the line that causes the issue: `document.getElementById('location-search-autocomplete').blur();`
4. It looks like coordinates aren't getting formatted quite right in polar projections:
![Screen Shot 2022-02-10 at 9 18 31 AM](https://user-images.githubusercontent.com/1480000/153450021-3bbc3c1e-0bf2-4736-938a-67623594f702.png)
5. There should be a comma between the lat, lon values as seen in the menu (seems to be there when copied to clipboard):
![Screen Shot 2022-02-10 at 9 21 59 AM](https://user-images.githubusercontent.com/1480000/153450788-c95d2634-34f3-4145-ae33-0ab40211bd51.png)

jasontk19 reviewed (changes requested):
Added a few code comments and found two bugs:

1. Right click
2. Copy coordinates
3. Right click again to bring up menu
4. Context menu gets removed when tooltip gets removed
---
1. Click "place location marker" button in location search tool
2. Right-click on the map to cancel placement
3. Error is thrown

jwhurley1 commented on web/js/components/context-menu/context-menu.js:
```diff
@@ -0,0 +1,164 @@
+    unitOfMeasure,
+    coordinates,
+    config,
+    getReverseGeocode: (coords) => reverseGeocode(coords, config),
```
Fixed in https://github.com/nasa-gibs/worldview/pull/3810/commits/0047ca032d96f8f0945d51c3802ed2d64df6a8ff

jwhurley1 commented on web/js/containers/map-interactions/map-interactions.js:
```diff
@@ -27,9 +29,13 @@ class MapInteractions extends PureComponent {
     const mapClasses = this.getMapClasses();
     return (
       <>
+        {/* <ContextMenu /> */}
```
Fixed in https://github.com/nasa-gibs/worldview/pull/3810/commits/0047ca032d96f8f0945d51c3802ed2d64df6a8ff

jasontk19 commented on web/js/components/context-menu/context-menu.js:
```diff
@@ -74,10 +70,13 @@ function RightClickMenu(props) {
   useEffect(() => {
     events.on('map:singleclick', handleClick);
     events.on('map:contextmenu', handleContextEvent);
+    if (isCoordinateSearchActive) {
```
It looks like this fixes the issue.  Alternatively, if the idea is that we don't need the listeners when `isCoordinateSearchActive` is `true`, we could skip registering them in the first place like:
```
useEffect(() => {
    if (isCoordinateSearchActive) return;
    events.on('map:singleclick', handleClick);
    events.on('map:contextmenu', handleContextEvent);
    return () => {
      events.off('map:singleclick', handleClick);
      events.off('map:contextmenu', handleContextEvent);
    };
  });
```

jwhurley1 replied:
Great suggestion! I think this is better solution and is a little more clear.  I'll put this change in and push it up.
